dvc/utils/__init__.py

```diff
@@ -62,7 +62,10 @@ def file_md5(fname):
                     if not data:
                         break
 
-                    if binary:
+                    if is_file_binary is None:
+                        is_file_binary = not istext(data)
```

Comment:
This basically kiils the heuristics and starts checking the whole file contents. I suppose we could extend the heuristics to "check first N bytes + check last N bytes" to cover your particular case, but there is no real need for it, as the main intention for this heuristics is providing md5 compatibility for git-tracked files between *nix and windows, where git uses CRLF by default even if you didn't have them before in your file. So we actually need to have a heuristic that is an exact match to the one that git uses. :slightly_smiling_face:

That being said, we sure could look into optimizing this by not opening the file again or optimizing dos2unix() etc.

Reply:
FWIW: This code still only checks the first N bytes. Where N was 512 bytes before, now it's the first MB.

Reply:
Looks like Git may use [8K](https://git.kernel.org/pub/scm/git/git.git/tree/xdiff-interface.c?id=HEAD#n187)
See also: https://stackoverflow.com/questions/6119956/how-to-determine-if-git-handles-a-file-as-binary-or-as-text